### PR TITLE
Bug 1780392: Fix `hasModel` check for CRD models

### PR DESCRIPTION
**Pull request by spadgett**
Fixes https://jira.coreos.com/browse/ODC-2125

/assign @andrewballantyne @christianvogt

### CODE REVIEW

**Comment by andrewballantyne** [comment]
You're fast :) But this will fail the e2e test as it did for me back in Oct.

image

**Comment by andrewballantyne** [comment]
/hold

I could be wrong, but I'll add the hold for now to prevent auto-merge.

**Comment by andrewballantyne** [comment]
/hold cancel

Didn't see the WIP :)

File: frontend/public/module/k8s/k8s-models.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ export const modelsToMap = (models: K8sKind[]): ImmutableMap<K8sResourceKindRefe
 let k8sModels = modelsToMap(_.values(staticModels));
 
 const hasModel = (model: K8sKind) =>
-  k8sModels.has(referenceForModel(model)) || k8sModels.has(model.kind);
+  k8sModels.has(referenceForModel(model)) || (!model.crd && k8sModels.has(model.kind));
```
Comment (andrewballantyne) [inline frontend/public/module/k8s/k8s-models.ts]: Now that you got it to work where I failed... I would recommend going with my first implementation and making `hasModel` and `modelsToMap` use the same logic to figure out the key to store it under.

From Slack:

> https://github.com/spadgett/console/blob/586f68d0eaa32defc34df2554b39c5fc226a7d79/frontend/public/module/k8s/k8s-models.ts#L13-L18
> I think we need to de-duplicate and reuse this logic. The hasModel should lookup in the same order/reference as the storage.

File: frontend/public/module/k8s/k8s-models.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@ export const modelsToMap = (models: K8sKind[]): ImmutableMap<K8sResourceKindRefe
 let k8sModels = modelsToMap(_.values(staticModels));
 
 const hasModel = (model: K8sKind) =>
```
Comment (andrewballantyne) [inline frontend/public/module/k8s/k8s-models.ts]: Suggested change:

```
const modelKey = (model) => {
  // TODO: Use `referenceForModel` even for known API objects
  return model.crd ? referenceForModel(model) : model.kind;
};

export const modelsToMap = (models: K8sKind[]): ImmutableMap<K8sResourceKindReference, K8sKind> => {
  return ImmutableMap<K8sResourceKindReference, K8sKind>().withMutations((map) => {
    models.forEach((model) => map.set(modelKey(model), model));
  });
};
...
const hasModel = (model: K8sKind) => k8sModels.has(modelKey(model));
```

That way they are both based off the same value, it looks it up in the same fashion it stores it. No way for it to be confused on reverse lookups.

File: frontend/public/module/k8s/k8s-models.ts (hunk 1 above)
Comment (andrewballantyne) [reply frontend/public/module/k8s/k8s-models.ts]: Suggested change:

```
const modelKey = (model) => {
  // TODO: Use `referenceForModel` even for known API objects
  return model.crd ? referenceForModel(model) : model.kind;
};

export const modelsToMap = (models: K8sKind[]): ImmutableMap<K8sResourceKindReference, K8sKind> => {
  return ImmutableMap<K8sResourceKindReference, K8sKind>().withMutations((map) => {
    models.forEach((model) => map.set(modelKey(model), model));
  });
};
...
const hasModel = (model: K8sKind) => k8sModels.has(modelKey(model));
```

That way they are both based off the same value, it looks it up in the same fashion it stores it. No way for it to be confused on reverse lookups.

Comment (spadgett) [reply frontend/public/module/k8s/k8s-models.ts]: Unfortunately, this changes the logic a bit and breaks when you pass the full reference for a model that has `crd: false`. For instance, `route.openshift.io~v1~Route` will no longer locate the route model.

Comment (spadgett) [reply frontend/public/module/k8s/k8s-models.ts]: Disregard, you're right

**Comment by christianvogt** [comment]
Such familiar code :)
we were sooo close lol

**Comment by spadgett** [comment]
/hold

Looks like I broke the operands list: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_console/3690/pull-ci-openshift-console-master-e2e-gcp-console/5521/artifacts/e2e-gcp-console/gui_test_screenshots/83e0defc188d63aa347f87370195c90d.png

**Comment by spadgett** [comment]
The OLM pages work fine clicking through the UI, but this breaks legacy URLs that used `clusterserviceversions` in the URL. Since the model is `crd: true`, `connectToPlural` doesn't find the model.

**Pull request merged**